### PR TITLE
Compute angle between two vectors

### PR DESCRIPTION
**Pull request by ezrayst**
- Add function in `geometry` to compute angle between two vectors
- Input is `VectorXd` so this is good for 3D and 2D
- Test cases are made ready

### CODE REVIEW

File: include/geometry.h [hunk 1]
```diff
@@ -23,6 +23,13 @@ class Geometry {
   Eigen::Matrix<double, Tdim, Tdim> inverse_rotation_matrix(
       const Eigen::Matrix<double, Tdim, 1>& angles) const;
 
+  //! Return angle between two vectors in radians
+  //! \param[in] vector_a First vector
+  //! \param[in] vecotr_b Second vector
+  //! \retval anti-clockwise angle from vector_a to vector_b
+  double angle_between_two_vectors(const Eigen::VectorXd& vector_a,
```
Comment (kks32) [inline include/geometry.h]: Why can't it be `Vector<double, Tdim, 1>`?

File: include/geometry.tcc [hunk 2]
```diff
@@ -45,4 +45,16 @@ inline Eigen::Matrix<double, 3, 3> mpm::Geometry<3>::inverse_rotation_matrix(con
...
+//! Return angle between two vectors in radians
+template <unsigned Tdim>
+double mpm::Geometry<Tdim>::angle_between_two_vectors(
+    const Eigen::VectorXd& vector_a, const Eigen::VectorXd& vector_b) {
+
+  // angle between vector a and b = arccos( a dot b / ||a|| ||b||)
+  const double angle =
+      acos((vector_a.normalized()).dot((vector_b.normalized())));
+
+  return angle;
```
Comment (kks32) [inline include/geometry.tcc]: Use a return value optimisation, and make this function inline

File: tests/geometry_test.cc [hunk 3]
```diff
@@ -17,6 +17,7 @@ TEST_CASE("Geometry is checked for 2D case", "[geometry][2D]") {
 
   SECTION("Check inverse rotation matrix") {
```
Comment (kks32) [inline tests/geometry_test.cc]: Remove empty line

File: tests/geometry_test.cc [hunk 4]
```diff
@@ -41,6 +42,21 @@ TEST_CASE("Geometry is checked for 2D case", "[geometry][2D]") {
...
+    // Make geometry
+    const auto geometry = std::make_unique<mpm::Geometry<Dim>>();
+
+    Eigen::Matrix<double, 2, 1> vector_a;
+    vector_a << 3, 0;
+    Eigen::Matrix<double, 2, 1> vector_b;
+    vector_b << -2, 2;
+    const double angle = 2.356194490192345;
+
+    auto check_angle = geometry->angle_between_two_vectors(vector_a, vector_b);
```
Comment (kks32) [inline tests/geometry_test.cc]: Remove unnecessary `check_angle`

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/geometry.h (hunk 1 above)
Comment (ezrayst) [reply include/geometry.h]: We can do that too

File: include/geometry.tcc [hunk 5]
```diff
@@ -45,4 +45,14 @@ inline Eigen::Matrix<double, 3, 3> mpm::Geometry<3>::inverse_rotation_matrix(con
 
   // Return inverted rotation matrix
   return rotation_matrix.inverse();
-}
\ No newline at end of file
+}
+
+//! Return angle between two vectors in radians
+template <unsigned Tdim>
+double mpm::Geometry<Tdim>::angle_between_two_vectors(
```
Comment (kks32) [inline include/geometry.tcc]: Please make this a const function and inline

File: tests/geometry_test.cc [hunk 6]
```diff
@@ -79,4 +94,19 @@ TEST_CASE("Geometry is checked for 3D case", "[geometry][3D]") {
       }
     }
   }
+
+  SECTION("Check angle between two vectors") {
+    // Make geometry
+    const auto geometry = std::make_unique<mpm::Geometry<Dim>>();
+
+    Eigen::Matrix<double, 3, 1> vector_a;
+    vector_a << 3, 0, 0;
```
Comment (kks32) [inline tests/geometry_test.cc]: Could you use non-zero third dimension? Otherwise it is the same as 2D

**Review by kks32 — CHANGES_REQUESTED** [review]
Why is the final commit called ` optimize initialization`? What part of initialisation is being optimised?

I don't understand the reasoning behind this function? What if the 2 vectors are not on the same plane?

File: tests/geometry_test.cc (hunk 6 above)
Comment (ezrayst) [reply tests/geometry_test.cc]: Oh, yes!

**Comment by ezrayst** [comment]
We need to get the Euler angles (our definition) to define new coordinate system. For that, we need to compute 2 angles for 2D and 3 angles for 3D.

For instance, the definition of `gamma` angle would be between vector [0 0 1] to the normal to the face. Once obtained, we can rotate coordinate system.

**Comment by kks32** [comment]
But this function returns only one angle, you mentioned you need to compute 2 angles for 2D and 3 for 3D?

**Comment by ezrayst** [comment]
As the title suggests, this function only computes an angle between two vectors. 2 angles for 2D would be done in other functions (in `cell`) that call this function.

**Comment by kks32** [comment]
I don't understand which angle you are computing if the two vectors are not on the same plane. In other words, where do you intend to use this function?

**Comment by ezrayst** [comment]
I can always get an angle between any two vectors. The angle from vector a to vector b and the angle vice versa should add up to 360. It is in geometry, so it should be general mathematics. Simply calculating an angle between two random vectors.

I intend to call this function to calculate angle `alpha`, `beta` and `gamma`. Once we get the normal to the face (the new `z` axis) and specify the new `x` axis, thus automatically get the new `y` axis (it has to be orthogonal to both `x` and `z`), then we can call this function to get all angles. Then we feed them to rotation matrix to transform velocity.

**Comment by kks32** [comment]
This basically computes the angle between two vectors `on a given plane`

File: include/geometry.h [hunk 7]
```diff
@@ -23,6 +23,14 @@ class Geometry {
   Eigen::Matrix<double, Tdim, Tdim> inverse_rotation_matrix(
       const Eigen::Matrix<double, Tdim, 1>& angles) const;
 
+  //! Return angle between two vectors in radians
+  //! \param[in] vector_a First vector
+  //! \param[in] vecotr_b Second vector
+  //! \retval anti-clockwise angle from vector_a to vector_b
+  const double angle_between_two_vectors(
```
Comment (kks32) [inline include/geometry.h]: Can we rename to `angle_between_vectors`

File: include/geometry.h [hunk 8]
```diff
@@ -23,6 +23,14 @@ class Geometry {
   Eigen::Matrix<double, Tdim, Tdim> inverse_rotation_matrix(
       const Eigen::Matrix<double, Tdim, 1>& angles) const;
 
+  //! Return angle between two vectors in radians
```
Comment (kks32) [inline include/geometry.h:26]: Here say that on the shared plane to make it more obvious?

File: include/geometry.h [hunk 9]
```diff
@@ -23,6 +23,14 @@ class Geometry {
...
+  //! \param[in] vector_a First vector
+  //! \param[in] vecotr_b Second vector
+  //! \retval anti-clockwise angle from vector_a to vector_b
```
Comment (kks32) [inline include/geometry.h:29]: There is no more `\retval`

File: include/geometry.tcc [hunk 10]
```diff
@@ -45,4 +45,24 @@ inline Eigen::Matrix<double, 3, 3> mpm::Geometry<3>::inverse_rotation_matrix(con
 
   // Return inverted rotation matrix
   return rotation_matrix.inverse();
+}
+
+//! Return angle between two vectors in radians
+template <>
+inline const double mpm::Geometry<2>::angle_between_two_vectors(
```
Comment (kks32) [inline include/geometry.tcc]: Why do we specialise when we have the same implementation?

File: tests/geometry_test.cc [hunk 11]
```diff
@@ -79,4 +94,19 @@ TEST_CASE("Geometry is checked for 3D case", "[geometry][3D]") {
       }
     }
   }
+
+  SECTION("Check angle between two vectors") {
```
Comment (kks32) [inline tests/geometry_test.cc:98]: add `on the shared plane`

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/geometry.h (hunk 9 above)
Comment (ezrayst) [reply include/geometry.h:29]: what do you mean there is no more? We don't use it or the function does not return value?

File: include/geometry.h (hunk 7 above)
Comment (ezrayst) [reply include/geometry.h]: I think it's good. We can't have angle between more than 2 vectors anyway

File: include/geometry.h (hunk 8 above)
Comment (ezrayst) [reply include/geometry.h:26]: I am fine in adding it, but seems unnecessary. If not on the shared plane, the angle is either 0 or 360 right?

**Review by kks32 — APPROVED** [review]

**Pull request merged**